File: backend/app/models/video_asset.py
from __future__ import annotations

from datetime import datetime
from typing import Any, Optional, Literal

from pydantic import BaseModel, Field


class VideoAsset(BaseModel):
    """Represents a stored video asset and associated metadata."""

    id: str = Field(..., description="Unique asset identifier")
    project_id: str = Field(..., description="Project the asset belongs to")
    filename: str = Field(..., description="Stored filename")
    original_filename: str = Field(..., description="Original uploaded filename")
    relative_path: str = Field(..., description="Path relative to storage root")
    checksum: str = Field(..., description="SHA-256 checksum of file contents")
    size_bytes: int = Field(..., ge=0, description="File size in bytes")
    mime_type: str = Field(..., description="MIME type of the uploaded file")
    category: Literal["source", "processed", "proxy", "export", "thumbnail", "music"] = Field(
        default="source",
        description="Logical category for the asset within the video pipeline",
    )
    status: str = Field(default="uploaded", description="Processing status of the asset")
    thumbnail_path: Optional[str] = Field(
        default=None,
        description="Relative path to generated thumbnail if available",
    )
    source_asset_ids: list[str] = Field(
        default_factory=list,
        description="Identifiers of source assets used to produce this derivative file",
    )
    metadata: dict[str, Any] = Field(default_factory=dict, description="Additional extracted metadata")
    created_at: datetime = Field(default_factory=datetime.utcnow)
    updated_at: datetime = Field(default_factory=datetime.utcnow)

    class Config:
        json_encoders = {datetime: lambda v: v.isoformat()}

File: frontend/cto/lib/src/data/models/video_asset.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'video_asset.freezed.dart';
part 'video_asset.g.dart';

enum VideoAssetCategory {
  source,
  processed,
  proxy,
  export,
  thumbnail,
  music,
}

@freezed
class VideoAsset with _$VideoAsset {
  const factory VideoAsset({
    required String id,
    @JsonKey(name: 'project_id') required String projectId,
    required String filename,
    @JsonKey(name: 'original_filename') required String originalFilename,
    @JsonKey(name: 'relative_path') required String relativePath,
    required String checksum,
    @JsonKey(name: 'size_bytes') required int sizeBytes,
    @JsonKey(name: 'mime_type') required String mimeType,
    @Default(VideoAssetCategory.source) VideoAssetCategory category,
    @Default('uploaded') String status,
    @JsonKey(name: 'thumbnail_path') String? thumbnailPath,
    @JsonKey(name: 'source_asset_ids') @Default([]) List<String> sourceAssetIds,
    @Default({}) Map<String, dynamic> metadata,
    @JsonKey(name: 'created_at') required DateTime createdAt,
    @JsonKey(name: 'updated_at') required DateTime updatedAt,
  }) = _VideoAsset;

  factory VideoAsset.fromJson(Map<String, dynamic> json) => _$VideoAssetFromJson(json);
}

@freezed
class VideoUploadResponse with _$VideoUploadResponse {
  const factory VideoUploadResponse({
    @JsonKey(name: 'asset_id') required String assetId,
    required String filename,
    @JsonKey(name: 'original_filename') required String originalFilename,
    @JsonKey(name: 'size_bytes') required int sizeBytes,
    @JsonKey(name: 'project_id') required String projectId,
    required String status,
    @Default('Upload successful') String message,
  }) = _VideoUploadResponse;

  factory VideoUploadResponse.fromJson(Map<String, dynamic> json) =>
      _$VideoUploadResponseFromJson(json);
}

@freezed
class StorageStats with _$StorageStats {
  const factory StorageStats({
    String? root,
    @JsonKey(name: 'project_id') String? projectId,
    required Map<String, Map<String, dynamic>> categories,
  }) = _StorageStats;

  factory StorageStats.fromJson(Map<String, dynamic> json) => _$StorageStatsFromJson(json);
}
